### PR TITLE
[NCL-7768] Implement Product Milestone Comparison table

### PR DESCRIPTION
**Pull request by patrikk0123**

### CODE REVIEW

**Review by matedo1 — CHANGES_REQUESTED** [review]

**Comment by matedo1** [comment]
Hi @patrikk0123 , service container changes are merged (#203), feel free to rebase this PR on top of it.

File: src/services/productMilestoneApi.tsx [hunk 1]
```diff
@@ -65,3 +65,17 @@ export const getCloseResults = ({ id }: IProductMilestoneApiData, requestConfig:
 export const getStatistics = ({ id }: IProductMilestoneApiData, requestConfig: AxiosRequestConfig = {}) => {
   return mockClient.getHttpClient().get<any>(`/product-milestones/${id}/statistics`, requestConfig);
 };
+
+/**
+ * Gets Product Milestone-Artifact table.
+ *
+ * * @param serviceData - object containing:
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/services/productMilestoneApi.tsx]: done

File: src/components/ServiceContainers/ServiceContainerLoading.tsx [hunk 2]
```diff
@@ -14,6 +14,7 @@ interface IServiceContainerLoadingProps extends IServiceContainerProps {
   loadingDelayMilliseconds?: number;
   hasSkeleton?: boolean;
   variant?: 'block' | 'inline' | 'icon';
+  initLoadingState?: boolean;
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ServiceContainers/ServiceContainerLoading.tsx]: done

File: src/components/ProductMilestoneComparisonTablePage/ProductMilestoneComparisonTablePage.tsx [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+  );
+
+  useTitle(PageTitles.productMilestoneComparisonTable);
+
+  return (
+    <PageLayout
+      title={PageTitles.productMilestoneComparisonTable}
+      description={
+        <>
+          Table lists Artifacts and their versions in selected Product Milestones. Product Milestones can be removed from the
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTablePage/ProductMilestoneComparisonTablePage.tsx]: @patrikk0123 @janinko , what's the current state?

File: src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    }
+  }, [areAllArtifactsExpanded, serviceContainerProductMilestoneComparisonTable.data?.content]);
+
+  return (
+    <>
+      <Toolbar>
+        <ToolbarItem>
+          <SearchSelect
+            selectedItem={selectedProduct?.name}
+            onSelect={(_, entity?: any) => {
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx]: done

File: src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+            fetchCallback={productApi.getProducts}
+            titleAttribute="name"
+            width={416}
+            placeholderText="Select Product"
+          />
+        </ToolbarItem>
+        <ToolbarItem>
+          <SearchSelect
+            selectedItem={selectedProductVersion?.version}
+            onSelect={(_, entity?: any) => {
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx]: done

File: src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx [hunk 6]
```diff
@@ -0,0 +1,356 @@
...
+            titleAttribute="version"
+            width={200}
+            placeholderText="Select Version"
+            isDisabled={!selectedProduct}
+          />
+        </ToolbarItem>
+        <ToolbarItem>
+          <SearchSelect
+            selectedItem={selectedProductMilestone?.version}
+            onSelect={(_, entity?: any) => setSelectedProductMilestone(entity)}
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx]: done

File: src/components/ProductMilestoneComparisonTablePage/ProductMilestoneComparisonTablePage.tsx [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+
+import { useServiceContainer } from 'hooks/useServiceContainer';
+import { useTitle } from 'hooks/useTitle';
+
+import { PageLayout } from 'components/PageLayout/PageLayout';
+import { ProductMilestoneComparisonTable } from 'components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable';
+
+import * as productMilestoneApi from 'services/productMilestoneApi';
+
+interface IProductMilestoneComparisonTablePageProps {
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTablePage/ProductMilestoneComparisonTablePage.tsx]: done

File: src/services/productMilestoneApi.tsx [hunk 8]
```diff
@@ -65,3 +65,17 @@ export const getCloseResults = ({ id }: IProductMilestoneApiData, requestConfig:
...
 };
+
+/**
+ * Gets Product Milestone Comparison table.
+ *
+ * * @param serviceData - object containing:
+ *  - data - array of Product Milestone ids
+ * @param requestConfig - Axios based request config
+ */
+export const getProductMilestoneComparisonTable = (
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/services/productMilestoneApi.tsx]: done

File: src/services/productMilestoneApi.tsx [hunk 9]
```diff
@@ -65,3 +65,17 @@ export const getCloseResults = ({ id }: IProductMilestoneApiData, requestConfig:
...
+  { data }: { data: { productMilestones: string[] } },
+  requestConfig: AxiosRequestConfig = {}
+) => {
+  return mockClient.getHttpClient().post<any>(`/product-milestone-comparison-table`, data, requestConfig);
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/services/productMilestoneApi.tsx]: done

File: src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx [hunk 10]
```diff
@@ -0,0 +1,366 @@
...
+                });
+              }}
+              isAriaDisabled={
+                !productMilestoneColumns.length ||
+                !!selectedProductMilestone ||
+                serviceContainerProductMilestoneComparisonTable.loading ||
+                !wereColumnsChanged
+              }
+            >
+              Fetch
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx:231]: done

File: src/common/constants.ts [hunk 11]
```diff
@@ -14,6 +14,7 @@ export const PageTitles = {
   groupBuilds: 'Group Builds',
   artifacts: 'Artifacts',
   scmRepositories: 'SCM Repositories',
+  productMilestoneComparisonTable: 'Product Milestone Comparison',
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/common/constants.ts]: done

File: src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx [hunk 12]
```diff
@@ -0,0 +1,366 @@
...
+      setExpandedArtifacts(
+        serviceContainerProductMilestoneComparisonTable.data?.content.map((artifact: any) => artifact.identifier)
+      );
+    } else if (areAllArtifactsExpanded === false) {
+      setExpandedArtifacts([]);
+    }
+  }, [areAllArtifactsExpanded, serviceContainerProductMilestoneComparisonTable.data?.content]);
+
+  const wasSelectedMilestoneAdded = productMilestoneColumns.some(
+    (milestoneColumn: any) => selectedProductMilestone?.id === milestoneColumn.id
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx]: Why also type `IProductMilestoneColumn` was not added?

File: src/components/SearchSelect/SearchSelect.tsx [hunk 13]
```diff
@@ -31,32 +38,42 @@ interface ISearchSelectProps {
...
   delayMilliseconds = 200,
-  pageSizeDefault = 20,
+  pageSizeDefault = 10,
+  width,
+  placeholderText = 'string | !string | s?ring | st*ng',
+  isDisabled,
 }: ISearchSelectProps) => {
-  // filtered data downloaded using callback
+  // data downloaded using fetchCallback
   const [currentData, setCurrentData] = useState<any[]>([]);
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/SearchSelect/SearchSelect.tsx]: done

File: src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+                serviceContainerProductMilestoneComparisonTableRunner({
+                  serviceData: {
+                    data: { productMilestones: productMilestoneColumns.map((milestoneColumn) => milestoneColumn.id) },
+                  },
+                }).then(() => {
+                  setHasFetchedAnything(true);
+                  setWereColumnsChanged(false);
+                });
+              }}
+              isAriaDisabled={
```
[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx:224]: ok, I found documentation covering this behavior, please add following link as comment: `https://www.patternfly.org/v4/components/tooltip/accessibility/#additional-considerations`

File: src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (patrikk0123) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx]: It's not needed for `.some` or '.map` if the collection has type determined.

[the remark this answers is not in the document]
Comment (matedo1) [reply src/components/ProductMilestoneComparisonTable/ProductMilestoneComparisonTable.tsx]: `any` type was removed, done

**Review by matedo1 — APPROVED** [review]

**Review by matedo1 — APPROVED** [review]

**Pull request merged**